### PR TITLE
Basic header implementation

### PR DESCRIPTION
**Pull request by meain**
Instead of #102 

- [x]  better colors
- [x] open to configure colors in config file

imgur

### CODE REVIEW

**Comment by meain** [comment]
I am not so sure if we have to assign a key to toggle the visibility of the header. We could just go with a command like `/header [show|hide]` or just `/header`.

We could also even consider to renaming it to status bar than a header which make more sense to the use case that you are suggesting. Mostly just a terminology change and a command name change to something like `/statusline [show|hide]`.

**Comment by osa1** [comment]
Both sound good to me. Let's avoid `[show|hide]` part and make `/statusline` toggle status lines of _all_ tabs.

**Comment by meain** [comment]
All that we discussed apart from the below ones are good. Feel free to merge it. I will look into adding the channel topic to status line later.

- [ ] Show channel topic ( and maybe modes )
- [ ] Might wanna come up with better colors

**Comment by osa1** [comment]
Could you also implement resizing? If showing multiple lines (instead of one line) when there isn't enough room is too hard you could just hide the status bar.

**Comment by meain** [comment]
Oh, didn't think of such a scenario, will add that.

**Comment by meain** [comment]
@osa1 Are we handling resize once tiny is started?

**Comment by osa1** [comment]
Yep, see the `resize()` methods. You can also try it by resizing your terminal.

**Comment by meain** [comment]
When I thought about it the only info needed was the notification and ignore status (for now). Now, the status line adapts to showing that based on the available width.

**Comment by meain** [comment]
Btw, I tried to do this thinking that the resize will trigger a re-render but didn't really work. 
Related (https://github.com/osa1/tiny/blob/master/src/tui/mod.rs#L347)?

```diff
diff --git a/src/tui/mod.rs b/src/tui/mod.rs
index 04fbe45..b7fbdc5 100644
--- a/src/tui/mod.rs
+++ b/src/tui/mod.rs
@@ -571,6 +571,9 @@ impl TUI {
             tab_left = tab_left_;
             num_visible = num_visible_;
         }
+
+        // redraw after resize
+        self.draw()
     }
 
     pub fn get_nicks(&self, serv_name: &str, chan_name: &str) -> Option<&Trie> {
```

**Comment by meain** [comment]
## Final version: 

* micro

* mini

* small

* normal
)

**Comment by meain** [comment]
look good? @osa1

**Comment by osa1** [comment]
@meain Yep, I'm busy with other stuff, will merge soon. Feel free to start working on showing the channel title/modes if you want. I will only make small changes on the patch so I don't think there will be a lot of conflicts.

Btw, what is this problem you mention

> Btw, I tried to do this thinking that the resize will trigger a re-render but didn't really work.

if that's still reproducible can you open a ticket so that we don't forget about it? I don't quite understand what it is so can't open it myself.

**Comment by osa1** [comment]
@meain would you mind making a few changes? Let me know if you don't have the time or not interested.

- [ ] Let's make the status line always take one line. If there isn't enough room to show it the we should not show it. (a similar logic is implemented for current nick, see messaging.rs and how `draw_curent_nick` is used, but feel free to implement it differently)

- [ ] The layout is not great. The server/channel name is abbreviated even when there's enough room. I think we should have this invariant: if there's enough room to show everything in all details, we should do it. Only otherwise we should start abbreviating stuff.

- [ ] Finally, let's not implement this as an overlay, and resize the other widget (`MessagingUI`) when we decide to draw the status line (that is, whe the user did `/statusline` and there is enough room on the screen to draw it).

Can you also format the new file using `rustfmt`? (please don't format old files yet, we should make one diff formatting everything later)

**Comment by meain** [comment]
Sure, will be a bit busy this week but I will finish it up.

File: src/tui/mod.rs [hunk 1]
```diff
@@ -1155,6 +1153,10 @@ impl TUI {
 
     pub fn toggle_statusline(&mut self) {
         self.show_statusline = !self.show_statusline;
+        let statusline_height = if self.statusline_visible && self.show_statusline { 1 } else { 0 };
```
[the remark this answers is not in the document]
Comment (osa1) [reply src/tui/mod.rs]: Maybe try `RUST_BACKTRACE=1 cargo run 2>stderr`. After the crash you'll see a file `stderr` with a stack trace in it. That should show where the panic occurs.

If it doesn't help I can give it a try later today.

[the remark this answers is not in the document]
Comment (meain) [reply src/tui/mod.rs]: Fixed

**Comment by meain** [comment]
@osa1 looks good?

**Comment by osa1** [comment]
Thanks! Will take a look today.

**Pull request closed without merging**